src/maud/data_model.py

```diff
@@ -401,32 +391,45 @@ class StanCodeSet:
 
     def __init__(
         self,
-        metabolite_codes: Dict[str, int],
-        mic_codes: Dict[str, int],
-        balanced_mic_codes: Dict[str, int],
-        unbalanced_mic_codes: Dict[str, int],
-        reaction_codes: Dict[str, int],
-        experiment_codes: Dict[str, int],
-        enzyme_codes: Dict[str, int],
-        drain_codes: Dict[str, int],
-        phos_enz_codes: Dict[str, int],
+        metabolite_codes: List[str],
+        mic_codes: List[str],
+        balanced_mic_codes: List[str],
+        unbalanced_mic_codes: List[str],
+        km_codes: List[str],
+        reaction_codes: List[str],
+        experiment_codes: List[str],
+        enzyme_codes: List[str],
+        drain_codes: List[str],
+        phos_enz_codes: List[str],
+        yconc_exp_codes: List[str],
+        yconc_mic_codes: List[str],
+        yflux_exp_codes: List[str],
+        yflux_reaction_codes: List[str],
+        yenz_exp_codes: List[str],
+        yenz_enz_codes: List[str],
+        ci_mic_codes: List[str],
+        ai_mic_codes: List[str],
+        aa_mic_codes: List[str],
```

Comment:
Are we able to add back the indices? I know it's not as nice having .keys() everywhere but it feels like I have to interact with the data in a dense rather than sparse way. For example, if I have metabolites that I want to set in the initial conditions, I first loop through the measurements to find the metabolite, experiment and value, but then I need to go through the list in order to find out where to insert them in the array. I also see this as being an issue with perturbations for example, where I only want to change one metabolite/enzyme and I have to loop through the entire list in order to find out where to insert it.

Reply:
I really think this way is nicer - the main thing I was thinking is that in this form the codes can be used in both python and Stan without needing to worry about whether the index starts with zero or one. When you need the dictionary you can get it with the function [`codify`](https://github.com/biosustain/Maud/blob/ca1eeb0708be00b0462dacd8e2eb8c1dd61fe5b1/src/maud/utils.py#L26) - I think I did this in a few places.

I think maybe your concern is more with the changes to the `KineticModel` class, so that e.g. `km.metabolites` is now a list of metabolites rather than a dictionary? I agree it would be nice to keep the ability to get-by-id that comes for free with dictionaries, avoiding annoying loop searches. The reason I changed it was to make sure the order stays the same unless explicitly changed: I was worried that with dictionaries the entries might unintentionally get sorted by id at some point, messing up the indexing. Could we do this change in a separate PR to make sure nothing goes wrong?

Reply:
I'm just making changes to the water stoichiometry and the initial conditions at the moment. Currently my model doesn't run using this branch and I believe that's just due to the initial conditions + ODE tolerances. But changing the initial concentrations should make it better